psh/date/date.c

```diff
@@ -0,0 +1,223 @@
+		return 1;
+	}
+
+	tinfo = localtime(&(tv->tv_sec));
+
+	/* passing `&format[1]` for omission of '+' at the beginning */
+	if (strftime(buf, sizeof(buf), &format[1], tinfo) == 0) {
+		fprintf(stderr, "date: '%s' expansion too long\n", format);
+		return 1;
+	}
+	printf("%s\n", buf);
+
+	return 0;
+}
+
+
+static int psh_date_get(const char *format)
+{
+	struct timeval tv;
+
+	time(&(tv.tv_sec));
+	return psh_date_print(&tv, format);
+}
+
+
+static int psh_date_convert(char *timestring, const char *format, struct timeval *tv)
+{
+	char *end;
+
+	/* FIXME: no strptime() implementation in libphoenix */
+	if (format != NULL) {
+		fprintf(stderr, "date: chosen option does not support FORMAT\n");
+		return 1;
+	}
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	tv->tv_usec = 0;
+	tv->tv_sec = strtoull(timestring, &end, 10);
+	if ((end == timestring) || (*end != '\0')) {
+		fprintf(stderr, "Unrecognized argument: %s\n", timestring);
+		return 1;
+	}
+	return 0;
+}
+
+
+static int psh_date_parse(char *timestring, const char *format)
+{
+	struct timeval tv;
+
+	if (timestring != NULL && timestring[0] != '@') {
+		fprintf(stderr, "date: invalid date '%s'\n", timestring);
+		return 1;
+	}
+
+	/* error message printed by psh_date_convert() */
+	if (psh_date_convert(&(timestring[1]), format, &tv) != 0) {
+		return 1;
+	}
+
+	return psh_date_print(&tv, NULL);
+}
+
+
+static int psh_date_set(char *timestring, const char *format)
+{
+	struct timeval tv;
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+
+	/* error message printed by psh_date_convert() */
+	if (psh_date_convert(timestring, format, &tv) != 0) {
+		return 1;
+	}
+
+	if (settimeofday(&tv, &tz) != 0) {
+		printf("date: time set failed: %s\n", strerror(errno));
+		return 1;
+	}
+
+	return psh_date_get(NULL);
+}
+
+
+static int psh_date(int argc, char **argv)
+{
+	int c, status = 1;
+	char *timestring = NULL, *format = NULL;
+	enum mode m = mode_print;
+
+	while ((c = getopt(argc, argv, "hs:d:")) != -1) {
+		switch (c) {
+			case 'h':
+				m = mode_help;
+				break;
+
+			case 's':
+				m = mode_set;
+				timestring = optarg;
+				break;
+
+			case 'd':
+				m = mode_parse;
+				timestring = optarg;
+				break;
+
+			default:
+				m = mode_skip;
+				break;
+		}
+	}
+
+	/* evaluate additional arguments/format */
+	if (m != mode_skip) {
+		if (argc > optind + 1) {
+			m = mode_skip;
+			printf("Unrecognized argument: %s\n", argv[optind + 1]);
+		}
+		else if (argc != optind) {
+			format = argv[optind];
+		}
+	}
+
+	/* run proper option */
+	switch (m) {
+		case mode_set:
+			status = psh_date_set(timestring, format);
+			break;
+
+		case mode_print:
+			status = psh_date_get(format);
+			break;
+
+		case mode_parse:
+			status = psh_date_parse(timestring, format);
+			break;
+
+		case mode_help:
+			psh_date_help(argv[0]);
+			status = 0;
+			break;
+
+		default:
+			/* everything set for default */
+			break;
+	}
+
+	if (status) {
+		psh_date_help(argv[0]);
```

Comment:
not sure I wan't to have help printed every time the date getting / setting fails, eg:
<img width="292" alt="image" src="https://user-images.githubusercontent.com/3939983/160101946-0bab4606-f327-4b9a-b4d0-c59f50605668.png">

Reply:
I think I overinterpreted your last note about "still doesn't print usage and exit with fail". Removed that help print if fail occurs.

Reply:
the usual case is to print help on unrecognized option (and exit with error) - not print help in case of any problems encountered during program execution, eg. `busybox date`:
![image](https://user-images.githubusercontent.com/3939983/160145272-c77360ce-d1ad-481e-957a-45fe42d7a45e.png)


Not a PR blocker, so giving approve